### ISSUE TITLE
Self deadlock issue in single threaded apps

### ISSUE THREAD

**Issue by munroesj52** [sphde/sphde#44]
Working on a new application I found that if i exited then restarted the application the application would deadlock on the restart. This seemed to be more likely if significant new data was added in the previous run (using sascompoundheap). Once this starts it repeats until the SAS region is reset and the data reloaded.

gdb attach and backtrace showed a write lock conflict with another process/thread ID. This was odd because the there was only one thread.

Using sasutil dump i could see that the conflicting process/thread was from the previous run. So we some on was leaving a dangling lock (without matching unlock).

So i rebuild sphde with --enable-debug. This show saslockl.cpp <SasLockList template> reporting:
         <error - unlock requested for an address that is not locked>
is reported at least once in the process that run before the process that deadlock.

Other lock trace entries show this is the point where the write_thread_lock_count goes to 2 for this entry and never returns to 0 after this error (the write_thread_lock_count cycles between 2 and 1).

debug continues ....

**Comment by munroesj52** [comment]
this error message is generated from file saslockl.cpp:
    SasLockList<Item, Key>::unlockNode(Key keyAddr)

looking at the code:
template<class Item, class Key>
void
SasLockList<Item, Key>::unlockNode(Key keyAddr)
{
#ifdef __SOMDebugPrint__
   fprintf (stderr, "%s\n", __FUNCTION__);
#endif

   SasLockListNode<Item, Key> * iterPtr;
   lock->write_lock();

   // Search the list for the specified key
   for (iterPtr = headPtr; iterPtr != NULL; iterPtr = iterPtr->next)
    {
      if (iterPtr->item->operator==(keyAddr))
         break;
    }
   if (iterPtr == NULL)
    {
#ifdef __SOMDebugPrint__
      fprintf (stderr, "%s: error - unlock requested for an address that "
               "is not locked\n", __FUNCTION__);
#endif
      return;
    }
   iterPtr->item->unlock();	// unlock item
   lock->unlock();		// unlock list
}

We see that the member function starts with a lock->write_lock () applied to the whole list.  The function  expects to find a matching lock item then unlock both the matching item and the master lock list.

But in the case of no matching item, the code just returns. this leaves the master lock list "locked". 

Clearly we need to add lock->unlock(); to the error patch. I have tested with the following change:

@@ -291,6 +291,7 @@
       fprintf (stderr, "%s: error - unlock requested for an address that "
                "is not locked\n", __FUNCTION__);
 #endif
+      lock->unlock();		// unlock list
       return;
     }
    iterPtr->item->unlock();	// unlock item

This change seems to eliminate the deadlock above will does not resolve the original problem with objects are unlocked without a proceeding lock. 

debug continues ...

**Comment by munroesj52** [comment]
used gdb to set a breakpoint on the statement printing the error message in SasLockList<Item, Key>::unlockNode(Key keyAddr). Then used backtrace to find the offending statement in SASCompoundHeapAlloc:
```
	  if (lastHeader != heapHeader)
	    SASUnlock (lastHeader);
```
Which would normally unlock the lastHeader (normally the SASCompoundHeap "extent" we just allocated from for expanding CompoundHeaps). The error condition implies that logic above this point previously unlocked the SASCompoundHeap "extent" corresponding to lastheader. This error can only occur for expanding CompoundHeaps, otherwise (lastHeader == heapHeader).

This function is tricky because it has to handle both expanding and non-expanding Compoundheaps and handle several special locking scenarios:

1. The first (or only) SASCompoundHeap is always locked for write (special case) as it is both use for allocation and contains the SASCompoundExpandList for a expanding CompoundHeap.

- An allocate that fails for lack of free space will allocate and add a SASCompoundHeap to the end of the ExpandList  of the containing (first) expanding SASCompoundHeap.
- The first SASCompoundHeap of an expanding heap always included in the first element ([0]) of the ExpandList.
- SASCompoundHeapAlloc favors the current last SASCompoundHeap (list->heap[list->count - 1]) in the ExpandList for allocation.

2. This last CompoundHeap entry is always locked for write (special case) before allocation for any expanding CompoundHeap. Unless the last CompoundHeap entry is full or exceeds the (expandBlock->loadFactor) limit.

3. If the current last CompoundHeap is (effectively) full then we need to scan the ExpandList in case some space has been recently freed within other contained SASCompoundHeap blocks. This is tricky because:

- We need to lock for write any CompoundHeap before we check for space or allocate from that heap.
- We want to include the first, but not the last, CompoundHeap from the ExpandList, in this scan.
- The first CompoundHeap is already locked so we special case skipping the (redundant) lock on element 0.
- We don't want to lock the whole (all elements in the) ExpandList if we don't have to. So when/if we find space in the list, we stop the scan early. We save the index of this element to use later to unlock CompoundHeap elements we have locked.

4. If we fail to find any free space in this scan we will allocate a new SASCompoundHeap block and add it to the end of the ExpandList.

- We can assume that no other process/thread can "see" this new SASCompoundHeap element because we are holding the write lock for the first (containing) CompoundHeap and the ExpandList. 
- So we don't need to lock this specific "new" CompoundHeap for the first allocation from it.

5. Either we found freespace in one of the existing (contained) CompoundHeap blocks or we allocated a new CompoundHeap block and now we can allocate the requested (SASSimpleHeap) space. For this we use SASCompoundHeapAllocInternal() for the appropriate CompoundHeap.

6. Now we need to release any write locks for CompoundHeap elements we have examined.

- Need to unlock any CompoundHeap elements between the first and last elements that we have examined. The search many have terminated (found space) early so we need to know which elements we actually examined and only unlock those.
- Need to unlock the last CompoundHeap element. But only if this is can expanding CompoundHeap.
- Need to unlock the first (containing) CompoundHeap and return the allocated sub heap.

I told you it was complicated!

So this code seems to be wrong under some of the special cases above:
`	  if (lastHeader != expandHeader)
	    {
	      for (i = 1; i <= last_lock; i++)
		{
		  SASUnlock (list->heap[i]);
		}
	    }
`
an first blush including the "last_lock" index is required for the "early exit" case where we found free space early in the list. But may be wrong for the case where we scan to whole list and have to allocate a need to allocate a new CompoundHeap element (in that case "i <= last_lock" would include last CompoundHeap element which we will unLock again later ...)

Thinking ...

**Comment by munroesj52** [comment]
As i suspected. the loop limit "i <= last_lock;" is correct if the scan of the ExpandList found free space between and excluding first and last. Including the i= last_lock is necessary because we set last_lock = i when we found the free space. 

But this wrong when scan does no find free space in existing contained CompoundHeaps and we extend (create a new CompoundHeap and add it tot he list). In this case we unlock the last CompoundHeap early because we initialize "last_lock = list->count - 1;" and "<=' includes the last entry.

So ... we can either change the unlock loop termination or how we initialize and update last_lock. Or both. For not i think the expanding case is most common to I suggest the following restructing:

```
diff -urN sphde/src/sascompoundheap.cpp sphde-master-1.4/src/sascompoundheap.cpp
--- sphde/src/sascompoundheap.cpp	2017-07-08 12:19:38.000000000 -0500
+++ sphde-master-1.4/src/sascompoundheap.cpp	2017-12-24 16:27:34.718396845 -0600
@@ -480,7 +480,7 @@
 		      < expandBlock->loadFactor)
 		    {
 		      expandHeader = expandBlock;
-		      last_lock = i;
+		      last_lock = i + 1;
 		      break;
 		    }
 		}
@@ -489,21 +489,23 @@
 		  expandHeader = (SASCompoundHeapHeader *)
 		    SASCompoundHeapExpandCreate (heap);
 		}
+	      /* Need to check again, in case the
+	       * SASCompoundHeapExpandCreate is needed but fails.  */
+	      if (expandHeader != NULL)
+		newHeap = SASCompoundHeapAllocInternal (expandHeader);
+              /* Now that we have allocated from a CompoundHeap, we can
+               * unlock any (other than first and last) from the
+               * ExpandList that we have accessed.  */
+	      for (i = 1; i < last_lock; i++)
+		{
+		  SASUnlock (list->heap[i]);
+		}
 	    }
 	  else
 	    {
-	      expandHeader = lastHeader;
+	      newHeap = SASCompoundHeapAllocInternal (lastHeader);
 	    }
-	  if (expandHeader != NULL)
-	    newHeap = SASCompoundHeapAllocInternal (expandHeader);
 
-	  if (lastHeader != expandHeader)
-	    {
-	      for (i = 1; i <= last_lock; i++)
-		{
-		  SASUnlock (list->heap[i]);
-		}
-	    }
 	  if (lastHeader != heapHeader)
 	    SASUnlock (lastHeader);
 	}
@@ -1363,7 +1365,7 @@
 		      < expandBlock->loadFactor)
 		    {
 		      expandHeader = expandBlock;
-		      last_lock = i;
+		      last_lock = i + 1;
 		      break;
 		    }
 		}
@@ -1372,21 +1374,23 @@
 		  expandHeader = (SASCompoundHeapHeader *)
 		    SASCompoundHeapExpandCreate (heap);
 		}
+	      /* Need to check again, in case the
+	       * SASCompoundHeapExpandCreate is needed but fails.  */
+	      if (expandHeader != NULL)
+		newHeap = SPHCompoundPCQAllocInternal (expandHeader);
+              /* Now that we have allocated from a CompoundHeap, we can
+               * unlock any (other than first and last) from the
+               * ExpandList that we have accessed.  */
+	      for (i = 1; i < last_lock; i++)
+		{
+		  SASUnlock (list->heap[i]);
+		}
 	    }
 	  else
 	    {
-	      expandHeader = lastHeader;
+	      newHeap = SPHCompoundPCQAllocInternal (lastHeader);
 	    }
-	  if (expandHeader != NULL)
-	    newHeap = SPHCompoundPCQAllocInternal (expandHeader);
 
-	  if (lastHeader != expandHeader)
-	    {
-	      for (i = 1; i <= last_lock; i++)
-		{
-		  SASUnlock (list->heap[i]);
-		}
-	    }
 	  if (lastHeader != heapHeader)
 	    SASUnlock (lastHeader);
 	}
```

**Comment by RajalakshmiS** [comment]
Merged PR#46
